### PR TITLE
Changed classifier to inspect whole chain of throwables and not just the root cause

### LINKED ISSUE #310: DefaultFaultClassifier classifies root cause but not throwable itself
From the code it looks like that `DefaultFaultClassifier` classifies root cause but not throwable itself.

### PR DESCRIPTION
**Pull request by whiskeysierra**
Fixes #310

### CODE REVIEW

File: riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java [hunk 1]
```diff
@@ -14,13 +14,12 @@ public DefaultFaultClassifier(final Predicate<Throwable> isTransient) {
...
     @Override
     public Throwable classify(final Throwable throwable) {
-        final Throwable root = Throwables.getRootCause(throwable);
-
-        if (isTransient.test(root)) {
-            return new TransientFaultException(throwable);
-        }
-
-        return throwable;
+        return Stream.iterate(throwable, Throwable::getCause)
```
Comment (AlexanderYastrebov) [inline riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java]: Cool! I did not know about iterate

File: riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java [hunk 2]
```diff
@@ -14,13 +14,12 @@ public DefaultFaultClassifier(final Predicate<Throwable> isTransient) {
...
+                .filter(Objects::nonNull)
+                .findFirst()
```
Comment (AlexanderYastrebov) [inline riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java]: Looks like `.findFirst()` and `.filter(isTransient)` should be swapped

Comment (whiskeysierra) [reply riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java]: That actually will break if a non-transient throwable is given without any cause.

File: riptide-faults/src/test/java/org/zalando/riptide/faults/FaultClassifierTest.java [hunk 3]
```diff
@@ -35,6 +35,13 @@ public void shouldNotClassifyGenericSSLHandShakeException() {
         assertNotTransient(new SSLHandshakeException("No hands, no cookies"));
     }
 
+    @Test
+    public void shouldClassifyAsTransientWithNonTransientRootCause() {
```
Comment (AlexanderYastrebov) [inline riptide-faults/src/test/java/org/zalando/riptide/faults/FaultClassifierTest.java]: I think we need two additional tests, beside testing root cause match:
1. test that verifies matching throwable itself
2. test that verifies intermediate cause (`new Exception("outer", new Exception("intermediate", new Exception("inner")))`)

Comment (whiskeysierra) [reply riptide-faults/src/test/java/org/zalando/riptide/faults/FaultClassifierTest.java]: > test that verifies matching throwable itself

See line 20.

File: riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java (hunk 2 above)
Comment (AlexanderYastrebov) [reply riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java]: Why? iterate return throwable itself as first element

Comment (whiskeysierra) [reply riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java]: Yes, but if it's not transient then the stream is empty and findFirst will throw.

Comment (AlexanderYastrebov) [reply riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java]: > Returns an Optional describing the first element of this stream, or an empty Optional if the stream is empty.

Comment (whiskeysierra) [reply riptide-faults/src/main/java/org/zalando/riptide/faults/DefaultFaultClassifier.java]: I think I misinterpreted an exception that I saw during tests.

**Comment by AlexanderYastrebov** [comment]
:+1:

**Pull request merged**